cdc: support azure (#20858)
pingcap/docs: tidb-cloud/setup-azure-self-hosted-kafka-private-link-service.md

Reviewer: Most generic content in this doc is similar with https://docs.pingcap.com/tidbcloud/setup-self-hosted-kafka-private-link-service/

---
title: Set Up Self-Hosted Kafka Private Link Service in Azure
summary: This document explains how to set up Private Link service for self-hosted Kafka in Azure and how to make it work with TiDB Cloud.
---

# Set Up Self-Hosted Kafka Private Link Service in Azure

This document describes how to set up Private Link service for self-hosted Kafka in Azure, and how to make it work with TiDB Cloud.

The mechanism works as follows:

1. The TiDB Cloud Vnet connects to the Kafka Vnet through private endpoints.
2. Kafka clients need to communicate directly to all Kafka brokers.
3. Each Kafka broker is mapped to a unique port of endpoints within the TiDB Cloud Vnet.
4. Leverage the Kafka bootstrap mechanism and Azure resources to achieve the mapping.

The following diagram shows the mechanism. 

> ![Connect to Azure Self-Hosted Kafka Private Link Service](/media/tidb-cloud/changefeed/connect-to-azure-self-hosted-kafka-privatelink-service.jpeg)

Reviewer: Please use an image in the .png format instead.

Author: <img width="629" alt="Screen Shot 2025-05-08 at 09 35 07" src="https://github.com/user-attachments/assets/8840319c-9322-4c2c-bfcd-e1e802bc281d" />
It seems we are using both png and jpeg now.

        2. In **Connectivity Method**, select **Private Link**.
    4. Note down the information of the TiDB Cloud Azure account in **Reminders before proceeding**. You will use it to authorize TiDB Cloud to access the Kafka Private Link service.
    5. Enter a unique **Kafka Advertised Listener Pattern** for your Kafka Private Link service.
        1. Input a unique random string. It can only include numbers or lowercase letters. You will use it to generate **Kafka Advertised Listener Pattern** later.
        2. Click **Check usage and generate** to check if the random string is unique and generate **Kafka Advertised Listener Pattern** that will be used to assemble the EXTERNAL advertised listener for Kafka brokers. 

Note down all the deployment information. You need to use it to configure your Kafka Private Link service later.

The following table shows an example of the deployment information.

| Information     | Value    | Note    | 
|--------|-----------------|---------------------------|
| Region    | Virginia (`eastus`)    |  N/A |
| Subscription of TiDB Cloud Azure Account | `99549169-6cee-4263-8491-924a3011ee31`     |    N/A  |

> | Kafka Advertised Listener Pattern  | The unique random string: `abc` <br/> Generated pattern: &lt;broker_id&gt;.abc.eastus.azure.3199745.tidbcloud.com:&lt;port&gt; |

Author: not sure if it will show correctly, please make a preview first

    - **Name**: `kafka-lb`
    - **SKU**: `Standard`
    - **Type**: `Internal`
    - **Tier**: `Regional`

3. In **Frontend IP configuration** tab, add frontend IP configuration

    - **Name**: `kafka-lb-ip`
    - **IP version**: `IPv4`
    - **Virtual network**: `kafka-pls-vnet`
    - **Subnet**: `brokers-subnet`
    - **Assignment**: `Dynamic`
    - **Availability zone**: `Zone-redundant`

4. In **Backend pools**, add 3 backend pools

    - Name: pool1; Backend Pool Configuration: NIC; IP configurations: broker-node-1
    - Name: pool2; Backend Pool Configuration: NIC; IP configurations: broker-node-2
    - Name: pool3; Backend Pool Configuration: NIC; IP configurations: broker-node-3

5. In **Inbound rules**, add three load balancing rules, and create a health probe for these three rules.

    1. Rule 1

        - **Name**: rule1 
        - **IP version**: IPv4
        - **Frontend IP address**: kafka-lb-ip
        - **Backend pool**: `pool1`
        - **Protocol**: `TCP`
        - **Port**: `9093`
        - **Backend port**: `39092`
        - **Health probe**:
            - **Name**: `kafka-lb-hp`
            - **Protocol**: `TCP`

>             - **Port**: `39092`

Author: leave it as default

    
    2. Rule 2
        - **Name**: `rule1` 
        - **IP version**: `IPv4`
        - **Frontend IP address**: `kafka-lb-ip`
        - **Backend pool**: `pool1`
        - **Protocol**: `TCP`
        - **Port**: `9094`
        - **Backend port**: `39092`
        - **Health probe**: `kafka-lb-hp`

    3. Rule 3
    
        - **Name**: `rule1` 
        - **IP version**: `IPv4`
        - **Frontend IP address**: `kafka-lb-ip`
        - **Backend pool**: `pool1`
        - **Protocol**: `TCP`
        - **Port**: `9095`
        - **Backend port**: `39092`
        - **Health probe**: `kafka-lb-hp`
    
6. Click **Review + create**. Verify the information.

7. Click **Create**.

### 2. Set up Private Link Service

1. Go to [Azure Portal > Home > Private Link Center > Private link services](https://portal.azure.com/#view/Microsoft_Azure_Network/PrivateLinkCenterBlade/~/privatelinkservices), click **Create** to create a Private Link service for the Kafka load balancer.

2. In the **Basic** tab, fill in the parameters as follows:
    - Select your "Subscription", "Resource group" and "Region".
    - **Name**: `kafka-pls`

> 3. In the **Outbound settings** tab, fill in the parameters as follows:

Reviewer: ```suggestion
3. In the **Outbound settings** tab, fill in the parameters as follows, and then click **Next : Access security >**.

```

    - **Load balancer**: `kafka-lb`
    - **Load balancer frontend IP address**: `kafka-lb-ip`
    - **Source NAT subnet**: `kafka-pls-vnet/brokers-subnet`

> 4. In the **Access security** tab
>     - For "Visibility", select "Restricted by subscription" or "Anyone with your alias".
>     - For "Auto-approval", add "Subscription of TiDB Cloud Azure Account" you got before.

Reviewer: ```suggestion
4. In the **Access security** tab, do the following:

    - For **Visibility**, select **Restricted by subscription** or **Anyone with your alias**.
    - For **Subscription-level access and auto-approval**, click **Add subscriptions** to add the subscription of TiDB Cloud Azure account you got in [Prerequisites](#prerequisites).
```

> 5. Click **Review + create**. Verify the information.

Reviewer: ```suggestion
5. Click **Next : Tags >**, and then click  **Next : Review + create >** to verify the information.
```

> 6. Click **Create**. When the operation is done, note down the Alias of the Private Link service.
> 

Reviewer: ```suggestion
6. Click **Create**. When the operation is done, note down the alias of the Private Link service for later use.
```

## Step 3. Connect from TiDB Cloud

1. Return to the [TiDB Cloud console](https://tidbcloud.com) to create a changefeed for the cluster to connect to the Kafka cluster by **Private Link**. For more information, see [Sink to Apache Kafka](/tidb-cloud/changefeed-sink-to-apache-kafka.md).

2. When you proceed to **Configure the changefeed target > Connectivity Method > Private Link**, fill in the following fields with corresponding values and other fields as needed.

>     - **Kafka Advertised Listener Pattern**: `abc`. It is the same as the unique random string you use to generate **Kafka Advertised Listener Pattern** in [Prerequisites](#prerequisites).

Reviewer: ```suggestion
    - **Kafka Advertised Listener Pattern**: the unique random string that you use to generate **Kafka Advertised Listener Pattern** in [Prerequisites](#prerequisites).
```

>     - **Alias of the Private Link Service**: the Alias of the Private Link service.

Reviewer: ```suggestion
    - **Alias of the Private Link Service**: the alias of the Private Link service that you got in [2. Set up Private Link Service](#2-set-up-private-link-service).
```

    - **Bootstrap Ports**: `9093,9094,9095`. 

3. Proceed with the steps in [Sink to Apache Kafka](/tidb-cloud/changefeed-sink-to-apache-kafka.md).

Now you have successfully finished the task.

## FAQ

### How to connect to the same Kafka Private Link service from two different TiDB Cloud projects?

If you have already followed this document to successfully set up the connection from the first project, you can connect to the same Kafka Private Link service from the second project as follows:

1. Follow instructions from the beginning of this document. 

2. When you proceed to [Step 1. Set up a Kafka cluster](#step-1-set-up-a-kafka-cluster), follow [Reconfigure a running Kafka cluster](#reconfigure-a-running-kafka-cluster) to create another group of EXTERNAL listeners and advertised listeners. You can name it as **EXTERNAL2**. Note that the port range of **EXTERNAL2** can overlap with the **EXTERNAL**.

> 3. After reconfiguring brokers, create a new load balancer and a new Private Link service 

Reviewer: ```suggestion
3. After reconfiguring brokers, create a new load balancer and a new Private Link service.
```

| Kafka Advertised Listener Pattern  | The unique random string: `abc` <br/> Generated pattern: &lt;broker_id&gt;.abc.eastus.azure.3199745.tidbcloud.com:&lt;port&gt; |

## Step 1. Set up a Kafka cluster

If you need to deploy a new cluster, follow the instructions in [Deploy a new Kafka cluster](#deploy-a-new-kafka-cluster).

If you need to expose an existing cluster, follow the instructions in [Reconfigure a running Kafka cluster](#reconfigure-a-running-kafka-cluster). 

### Deploy a new Kafka cluster

#### 1. Set up the Kafka virtual network

1. Go to [Azure Portal > Home > Virtual networks](https://portal.azure.com/#browse/Microsoft.Network%2FvirtualNetworks), click **Create**.
2. In **Basics** tab
    1. Select your "Subscription", "Resource group" and "Region".
    2. Enter "Virtual network name", for example `kafka-pls-vnet`.
3. In **Security** tab, enable Azure Bastion
4. In **IP addresses** tab, do the following:
    1. Set address space, for example, `10.0.0.0/16`.
    2. Create a subnet for brokers:
        - **Name**: brokers-subnet
        - **IP address range**: 10.0.0.0 - 10.0.0.255
        - **Size**: /24

>     3. An AzureBastionSubnet will be created by default.

Author: 没关系， 这个可能是azure 的 UI 交互有调整吧， 整体就是填 ip 段。只要用户明白是这个网段就好 10.0.0.0/24

5. Click **Review + create**. Verify the information.
6. Click **Create**.

#### 2. Set up Kafka brokers

**2.1. Create broker nodes**

1. Go to [Azure Portal > Home > Virtual machines](https://portal.azure.com/#browse/Microsoft.Compute%2FVirtualMachines), click **Create**, and select **Azure virtual machine**.
2. In the **Basics** tab, fill parameters as following:
    - Select your "Subscription", "Resource group" and "Region".
    - **Virtual machine name**: `broker-node`
    - **Availability options**: `Availability zone`
    - **Zone options**: `Self-selected zone`
    - **Availability zone**: `Zone 1, Zone 2, Zone 3`
    - **Image**: `Ubuntu Server 24.04 LTS - x64 Gen2`
    - **VM architecture:** `x64`
    - **Size**: `Standard_D2s_v3` 
    - **Authentication type**: `SSH public key`
    - **Username**: azureuser
    - **SSH public key source:** `Generate new key pair`
    - **Key pair name**: `kafka_broker_key`
    - **Public inbound ports**: `Allow selected ports`
    - **Select inbound ports**: `SSH (22)`
3. In the **Networking** tab, fill in the parameters as following:
    - **Virtual network**: `kafka-pls-vnet`

>     - **Subnet**: `brokers-subnet`

Author: Don't have change it to exactly as UI display, such display may be changed. 'brokers-subnet' has enough information to refer to the subnet.

**2.2. Prepare Kafka runtime binaries**

1. Navigate to the page of every broker node "Home > {resource group} > broker-node-{1,2,3}".

2. Click **Connect > Connect via Bastion**, fill in the parameters as following:

    - **Authentication Type**: `SSH Private Key from Local File`
    - **Username**: azureuser
    - **Local File**: select the private key downloaded by browser before
    - Click "Open in new browser tab".
    
3. Click **Connect**. A new browser tab with a Linux terminal is displayed. You need to open three Linux terminals for the three broker nodes.

4. Download binaries in each broker node.

    ```shell
    # Download Kafka and OpenJDK, and then extract the files. You can choose the binary version based on your preference.
    wget https://archive.apache.org/dist/kafka/3.7.1/kafka_2.13-3.7.1.tgz
    tar -zxf kafka_2.13-3.7.1.tgz
    wget https://download.java.net/java/GA/jdk22.0.2/c9ecb94cd31b495da20a27d4581645e8/9/GPL/openjdk-22.0.2_linux-x64_bin.tar.gz
    tar -zxf openjdk-22.0.2_linux-x64_bin.tar.gz
    ```

**2.3. Set up Kafka nodes on each broker node**

> 1. Set up a KRaft Kafka cluster with three nodes. Each node acts as a broker and controller roles. For every broker:

Reviewer: ```suggestion
1. Set up a KRaft Kafka cluster with three nodes. Each node serves both as a broker and a controller. For each broker node:
```

>     1. For `listeners`, all three brokers are the same and act as brokers and controller roles:
>         1. Configure the same CONTROLLER listener for all **controller** role nodes. If you only want to add the **broker** role nodes, you do not need the CONTROLLER listener in `server.properties`.
>         2. Configure two **broker** listeners. INTERNAL for internal access, and EXTERNAL for external access from TiDB Cloud.

Reviewer: ```suggestion
    1. Configure `listeners`. All three brokers are the same and act as brokers and controller roles.
        1. Configure the same CONTROLLER listener for all **controller** role nodes. If you only want to add the broker role nodes, you can omit the CONTROLLER listener in `server.properties`.
        2. Configure two broker listeners: **INTERNAL** for internal Kafka client access and **EXTERNAL** for access from TiDB Cloud.
```

    
    2. For `advertised.listeners`, do the following:
        1. Configure an INTERNAL advertised listener for each broker using the internal IP address of the broker node, which allows internal Kafka clients to connect to the broker via the advertised address.

>         2. Configure an EXTERNAL advertised listener based on **Kafka Advertised Listener Pattern** you get from TiDB Cloud for every broker node to help TiDB Cloud differentiate between different brokers. Different EXTERNAL advertised listeners help Kafka clients from TiDB Cloud side route requests to the right broker.

Reviewer: ```suggestion
        2. Configure an EXTERNAL advertised listener based on **Kafka Advertised Listener Pattern** you get from TiDB Cloud for each broker node to help TiDB Cloud distinguish between different brokers. Different EXTERNAL advertised listeners help Kafka clients from the TiDB Cloud side route requests to the right broker.
```

>             - `<port>` differentiates brokers from Kafka Private Link service access points. Plan a port range for EXTERNAL advertised listeners of all brokers. These ports do not have to be actual ports listened to by brokers. They are ports listened to by the load balancer for Private Link service that will forward requests to different brokers.

Reviewer: ```suggestion
            - Use different `<port>` values to differentiate brokers in Kafka Private Link service access. Plan a port range for the EXTERNAL advertised listeners of all brokers. These ports do not have to be actual ports listened to by brokers. They are ports listened to by the load balancer in the Private Link service that will forward requests to different brokers.
```

    # 2. Configure EXTERNAL in "advertised.listeners" based on the "Kafka Advertised Listener Pattern" in the "Prerequisites" section.
    # 2.1 The pattern is "<broker_id>.abc.eastus.azure.3199745.tidbcloud.com:<port>".
    # 2.2 So the EXTERNAL can be "abc.eastus.azure.3199745.tidbcloud.com:9095". Replace <broker_id> with "b" prefix plus "node.id" properties, and replace <port> with a unique port (9095) in the port range of the EXTERNAL advertised listener.
    process.roles=broker,controller
    node.id=3
    controller.quorum.voters=1@{broker-node-1-ip}:29092,2@{broker-node-2-ip}:29092,3@{broker-node-3-ip}:29092
    listeners=INTERNAL://0.0.0.0:9092,CONTROLLER://0.0.0.0:29092,EXTERNAL://0.0.0.0:39092
    inter.broker.listener.name=INTERNAL
    advertised.listeners=INTERNAL://{broker-node-3-ip}:9092,EXTERNAL://b3.abc.eastus.azure.3199745.tidbcloud.com:9095
    controller.listener.names=CONTROLLER
    listener.security.protocol.map=INTERNAL:PLAINTEXT,CONTROLLER:PLAINTEXT,EXTERNAL:PLAINTEXT,SSL:SSL,SASL_PLAINTEXT:SASL_PLAINTEXT,SASL_SSL:SASL_SSL
    log.dirs=./data
    ```

3. Create a script, and then execute it to start the Kafka broker in each broker node.

    ```shell

> 

Reviewer: ```suggestion
```

    rm -rf $KAFKA_LOG_DIR
    mkdir -p $KAFKA_LOG_DIR

    $KAFKA_STORAGE_CMD format -t "BRl69zcmTFmiPaoaANybiw" -c "$KAFKA_CONFIG_DIR/server.properties" > $KAFKA_LOG_DIR/server_format.log   
    LOG_DIR=$KAFKA_LOG_DIR nohup $KAFKA_START_CMD "$KAFKA_CONFIG_DIR/server.properties" &
    ```

**2.4. Test the cluster setting**

1. Test the Kafka bootstrap.

    ```shell
    export JAVA_HOME=/home/azureuser/jdk-22.0.2

    # Bootstrap from INTERNAL listener
    ./kafka_2.13-3.7.1/bin/kafka-broker-api-versions.sh --bootstrap-server {one_of_broker_ip}:9092 | grep 9092
    # Expected output (the actual order might be different)
    {broker-node-1-ip}:9092 (id: 1 rack: null) -> (
    {broker-node-2-ip}:9092 (id: 2 rack: null) -> (
    {broker-node-3-ip}:9092 (id: 3 rack: null) -> (

    # Bootstrap from EXTERNAL listener
    ./kafka_2.13-3.7.1/bin/kafka-broker-api-versions.sh --bootstrap-server {one_of_broker_ip}:39092
    # Expected output for the last 3 lines (the actual order might be different)
    # The difference in the output from "bootstrap from INTERNAL listener" is that exceptions or errors might occur because advertised listeners cannot be resolved in kafka-pls-vnet.

>     # TiDB Cloud will make them resolvable and make it route to the right broker when you create a changefeed connect to this Kafka cluster by Private Link Service. 

Reviewer: ```suggestion
    # TiDB Cloud will make these addresses resolvable and route requests to the correct broker when you create a changefeed connected to this Kafka cluster via Private Link Service.
```

    create_topic() {
        echo "Creating topic if it does not exist..."
        $KAFKA_DIR/kafka-topics.sh --create --topic $TOPIC --bootstrap-server $BROKER_LIST --if-not-exists --partitions 3 --replication-factor 3
    }

    produce_messages() {
        echo "Producing messages to the topic..."
        for ((chrono=1; chrono <= 10; chrono++)); do
            message="Test message "$chrono
            echo "Create "$message
            echo $message | $KAFKA_DIR/kafka-console-producer.sh --broker-list $BROKER_LIST --topic $TOPIC
        done
    }
    create_topic
    produce_messages 
    ```

3. Create a consumer script `consume.sh` in the bastion node.

    ```shell
    BROKER_LIST=$1
    SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
    export JAVA_HOME="$SCRIPT_DIR/jdk-22.0.2"
    KAFKA_DIR="$SCRIPT_DIR/kafka_2.13-3.7.1/bin"
    TOPIC="test-topic"
    CONSUMER_GROUP="test-group"
    consume_messages() {
        echo "Consuming messages from the topic..."
        $KAFKA_DIR/kafka-console-consumer.sh --bootstrap-server $BROKER_LIST --topic $TOPIC --from-beginning --timeout-ms 5000 --consumer-property group.id=$CONSUMER_GROUP
    }
    consume_messages
    ```

> 4. Execute `produce.sh` and `consume.sh` to verify that the Kafka cluster is running. These scripts will also be reused for later network connection testing. The script will create a topic with `--partitions 3 --replication-factor 3`. Ensure that all these three brokers contain data. Ensure that the script will connect to all three brokers to guarantee that network connection will be tested.

Reviewer: ```suggestion
4. Execute `produce.sh` and `consume.sh` to verify that the Kafka cluster is running. These scripts will also be reused later to test network connectivity. The `produce.sh` script will create a topic with `--partitions 3 --replication-factor 3`. Ensure that all these three brokers contain data. Ensure that the script will connect to all three brokers to guarantee that network connection will be tested.
```

> 

Reviewer: What does "Ensure that the script will connect to all three brokers to guarantee that network connection will be tested." mean?  What can users do to ensure the script will connect to all three brokers?

Author: user just execute produce.sh script, produce.sh will help finish the test

     # Configure the EXTERNAL map
    listener.security.protocol.map=...,EXTERNAL:PLAINTEXT
     ```

3. After you reconfigure all the brokers, restart your Kafka brokers one by one.

**2. Test EXTERNAL listener settings in your internal network**

You can download Kafka and OpenJDK in your Kafka client node.

```shell
# Download Kafka and OpenJDK, and then extract the files. You can choose the binary version based on your preference.
wget https://archive.apache.org/dist/kafka/3.7.1/kafka_2.13-3.7.1.tgz
tar -zxf kafka_2.13-3.7.1.tgz
wget https://download.java.net/java/GA/jdk22.0.2/c9ecb94cd31b495da20a27d4581645e8/9/GPL/openjdk-22.0.2_linux-x64_bin.tar.gz
tar -zxf openjdk-22.0.2_linux-x64_bin.tar.gz
```

Execute the following script to test if the bootstrap works as expected.

```shell
export JAVA_HOME=~/jdk-22.0.2

# Bootstrap from the EXTERNAL listener
./kafka_2.13-3.7.1/bin/kafka-broker-api-versions.sh --bootstrap-server {one_of_broker_ip}:39092

# Expected output for the last 3 lines (the actual order might be different)
# There will be some exceptions or errors because advertised listeners cannot be resolved in your Kafka network. 

> # TiDB Cloud will make them resolvable and make it route to the right broker when you create a changefeed connect to this Kafka cluster by Private Link service. 

Reviewer: ```suggestion
# TiDB Cloud will make these addresses resolvable and route requests to the correct broker when you create a changefeed connected to this Kafka cluster via Private Link Service.
```

    2. In the left navigation pane, click **Changefeed**.
    3. On the **Changefeed** page, click **Create Changefeed** in the upper-right corner.
        1. In **Destination**, select **Kafka**.
        2. In **Connectivity Method**, select **Private Link**.
    4. Note down the region information and the subscription of the TiDB Cloud Azure account in **Reminders before proceeding**. You will use it to authorize TiDB Cloud to access the Kafka Private Link service.
    5. Enter a unique **Kafka Advertised Listener Pattern** for your Kafka Private Link service.
        1. Input a unique random string. It can only include numbers or lowercase letters. You will use it to generate **Kafka Advertised Listener Pattern** later.
        2. Click **Check usage and generate** to check if the random string is unique and generate **Kafka Advertised Listener Pattern** that will be used to assemble the EXTERNAL advertised listener for Kafka brokers. 

Note down all the deployment information. You need to use it to configure your Kafka Private Link service later.

> The following table shows an example of the deployment information.

Reviewer: preview of this table: https://pr.pingcap-docsite-preview.pages.dev/tidbcloud/setup-azure-self-hosted-kafka-private-link-service/#:~:text=The%20following%20table%20shows%20an%20example%20of%20the%20deployment%20information.

## Step 1. Set up a Kafka cluster

If you need to deploy a new cluster, follow the instructions in [Deploy a new Kafka cluster](#deploy-a-new-kafka-cluster).

If you need to expose an existing cluster, follow the instructions in [Reconfigure a running Kafka cluster](#reconfigure-a-running-kafka-cluster). 

### Deploy a new Kafka cluster

#### 1. Set up the Kafka virtual network

1. Log in to the [Azure portal](https://portal.azure.com/), go to the [Virtual networks](https://portal.azure.com/#browse/Microsoft.Network%2FvirtualNetworks) page, and then click **+ Create** to create a virtual network.
2. In the **Basic** tab, select your **Subscription**, **Resource group**, and **Region**, enter a name (for example, `kafka-pls-vnet`) in the **Virtual network name** field, and then click **Next**.
3. In the **Security** tab, enable Azure Bastion, and then click **Next**.
4. In the **IP addresses** tab, do the following:

    1. Set the address space of your virtual network, for example, `10.0.0.0/16`.
    2. Click **Add a subnet** to create a subnet for brokers, fill in the following information, and then click **Add**.
        - **Name**: `brokers-subnet`

>         - **IP address range**: `10.0.0.0 - 10.0.0.255`

Reviewer: In "IP address range", if we choose `10.0.0.0/16`, the corresponding value is `10.0.0.0 - 10.0.255.255`.
![image](https://github.com/user-attachments/assets/aec71cbe-7d78-468e-9838-401226e26104)

>         - **IP address range**: `10.0.0.0 - 10.0.0.255`

Reviewer: ```suggestion
        - **IP address range**: `10.0.0.0/24`
```

    create_topic() {
        echo "Creating topic if it does not exist..."
        $KAFKA_DIR/kafka-topics.sh --create --topic $TOPIC --bootstrap-server $BROKER_LIST --if-not-exists --partitions 3 --replication-factor 3
    }

    produce_messages() {
        echo "Producing messages to the topic..."
        for ((chrono=1; chrono <= 10; chrono++)); do
            message="Test message "$chrono
            echo "Create "$message
            echo $message | $KAFKA_DIR/kafka-console-producer.sh --broker-list $BROKER_LIST --topic $TOPIC
        done
    }
    create_topic
    produce_messages
    ```

3. Create a consumer script `consume.sh` in the bastion node.

    ```shell
    BROKER_LIST=$1
    SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
    export JAVA_HOME="$SCRIPT_DIR/jdk-22.0.2"
    KAFKA_DIR="$SCRIPT_DIR/kafka_2.13-3.7.1/bin"
    TOPIC="test-topic"
    CONSUMER_GROUP="test-group"
    consume_messages() {
        echo "Consuming messages from the topic..."
        $KAFKA_DIR/kafka-console-consumer.sh --bootstrap-server $BROKER_LIST --topic $TOPIC --from-beginning --timeout-ms 5000 --consumer-property group.id=$CONSUMER_GROUP
    }
    consume_messages
    ```

> 4. Execute `produce.sh` and `consume.sh` to verify that the Kafka cluster is running. These scripts will also be reused later to test network connectivity. The `produce.sh` script will create a topic with `--partitions 3 --replication-factor 3`. Ensure that all these three brokers contain data. Ensure that the script will connect to all three brokers to guarantee that network connection will be tested.

Reviewer: ```suggestion
4. Run the `produce.sh` and `consume.sh` scripts. These scripts automatically test connectivity and message flow to verify that the Kafka cluster is functioning correctly. The `produce.sh` script creates a topic with `--partitions 3 --replication-factor 3`, sends test messages, and connects to all three brokers using the `--broker-list` parameter. The `consume.sh` script reads messages from the topic to confirm successful message delivery.
```
